Fix: MQTTAbstractionImpl: flush/sync MQTT Connect message immediately

barsnick opened the pull request:
Previously, a socket was opened, but the MQTT Connect command message was not flushed to the broker until init() was finished. This could be critical if a module's init() phase waits for some reason, as the MQTT broker's default socket timeout might disconnect the client.

This behavior was observed with the mosquitto broker, where unregistered sockets were consistently disconnected after 94 seconds.

Insert an additional mqtt_sync() directly after mqtt_connect(), while the mqtt_sync() thread is not yet started.

Co-authored-by: Fabian Hartung <fabian.hartung@chargebyte.com>

barsnick committed:
Fix: MQTTAbstractionImpl: flush/sync MQTT Connect message immediately

Previously, a socket was opened, but the MQTT Connect command message was
not flushed to the broker until init() was finished. This could be critical
if a module's init() phase waits for some reason, as the MQTT broker's
default socket timeout might disconnect the client.

This behavior was observed with the mosquitto broker, where unregistered
sockets were consistently disconnected after 94 seconds.

Insert an additional mqtt_sync() directly after mqtt_connect(), while the
mqtt_sync() thread is not yet started.

Co-authored-by: Fabian Hartung <fabian.hartung@chargebyte.com>
Signed-off-by: Moritz Barsnick <moritz.barsnick@chargebyte.com>

FaHaGit committed:
Remove on_mqtt_disconnect() call before connecting to the MQTT broker

Signed-off-by: Fabian Hartung <fabian.hartung@chargebyte.com>

a-w50 commented on lib/mqtt_abstraction_impl.cpp:
```diff
@@ -486,6 +486,12 @@ bool MQTTAbstractionImpl::connectBroker(std::string& socket_path) {
+            EVLOG_error << fmt::format("Error during MQTT sync: {}", mqtt_error_str(error));
+            return false;
+        }
+
```
The original code wasn't using early returns, but here might be a good place to change that.  Something like:
```
if (mqtt_connect(..) != MQTT_OK) {
  return false;
}

if ((const auto error = mqtt_sync(...)) != MQTT_OK) {
  EVLOG...;
  return false;
}

on_mqtt_connect();
return true
```

More readable imho.

As was mentioned in the comments already, a possible better solution would be to start the _event loop_/_the polling thread_ directly after doing the connect - because why would you create a socket and then not use it for some time?  It is indeed interesting to see that the mqttc implementation doesn't send anything on the `mqtt_connect` command - probably due to easier separation for error handling.  Good find @FaHaGit

FaHaGit committed:
Integrate early returns

This commit integrates early returns in the mqtt_abstraction_impl.cpp
file. This is done to improve the readability of the code and to reduce
the nesting of the code.

Signed-off-by: Fabian Hartung <fabian.hartung@chargebyte.com>

FaHaGit committed:
Merge branch 'main' into cb/sync_mqtt_connect
